### PR TITLE
Adds onClose callback to AutoComplete

### PR DESCRIPTION
**Pull request by darrenhebner**
📝I'm migrating this PR over from the old repo.

### WHY are these changes introduced?
We have a use case where we want to clear the search query for the autocomplete after the merchant is finished interacting with the autocomplete. We will use this `onClose` callback to clear the query.

### WHAT is this pull request doing?
This PR makes changes to `ComboBox` where `onClose` is called every time the popover is closed.

## 

### How to 🎩

🖥 [Local development instructions](https://github.com/Shopify/polaris-react/blob/master/README.md#development)
🗒 [General tophatting guidelines](https://vault.shopify.com/developers/Tophatting)
📄 [Changelog guidelines](https://github.com/Shopify/polaris-react/blob/master/documentation/Versioning%20and%20changelog.md)

Open the console and check that `closed` is being logged whenever you close the popover.

<details>
<summary>Copy-paste this code in <code>playground/Playground.tsx</code>:</summary>

```jsx
/* eslint-disable */

import * as React from 'react';
import {Page, AppProvider, Stack, Tag, TextContainer} from '@shopify/polaris';
import Autocomplete from '../src/components/Autocomplete';
import _ from 'lodash';

interface Props {}
interface State {}

export default class DemoTagsAutocomplete extends React.Component<
  Props,
  State
> {
  // Quick Config
  ALLOW_MULTIPLE = true;
  RESULT_PAGE_LENGTH = 10;

  options = [
    {value: 'byward_market', label: 'Byward Market'},
    {value: 'cheese_pizza', label: 'Cheese Pizza'},
    {value: 'macaroni_pizza', label: 'Macaroni Pizza'},
    {value: 'pepperoni_pizza', label: 'Pepperoni Pizza'},
    {value: 'centretown', label: 'Centretown'},
    {value: 'hintonburg', label: 'Hintonburg'},
    {value: 'pineapple', label: 'Pineapple'},
    {value: 'pineapple_pizza', label: 'Pineapple Pizza'},
    {value: 'chinatown', label: 'Chinatown'},
    {value: 'kraft_singles', label: 'Kraft Singles'},
    {value: 'dasani_water', label: 'Dasani Water'},
    {value: 'ice_soup', label: 'Ice Soup'},
    {value: 'chana_masala', label: 'Chana Masala'},
    {value: 'eggs', label: 'Eggs'},
    {value: 'mango', label: 'Mango'},
    {value: 'butter_chicken', label: 'Butter Chicken'},
    {value: 'deprived_snakes', label: 'Deprived Snakes'},
    {value: 'pizza_pockets', label: 'Pizza Pockets'},
  ];

  state = {
    selected: [],
    inputText: '',
    options: this.options.slice(0, this.RESULT_PAGE_LENGTH),
    filteredOptions: this.options.slice(
      this.RESULT_PAGE_LENGTH,
      this.options.length - 1,
    ),
    loading: false,
  };

  updateText = (newValue: string) => {
    this.setState({inputText: newValue});
    this.filterAndUpdateOptions(newValue);
  };

  clearText = () => {
    const {inputText} = this.state;
    if (inputText == '') {
      return;
    }
    this.setState({inputText: ''});
  };

  removeTag = (tag: string) => {
    let newSelected: string[] = this.state.selected;
    newSelected.splice(newSelected.indexOf(tag), 1);
    this.setState({selected: newSelected});
  };

  addTag = (tag: string) => {
    if (!tag || tag === '') {
      return;
    }
    let tagLabel = '';
    tagLabel = tag.replace('_', ' ');
    tagLabel = _.startCase(tagLabel);

    this.options.push({value: tag, label: tagLabel});
    this.filterAndUpdateOptions(tag);
  };

  renderTags = (): React.ReactNode => {
    return this.state.selected.map((option: string) => {
      let tagLabel = '';
      tagLabel = option.replace('_', ' ');
      tagLabel = _.startCase(tagLabel);
      return (
        <Tag key={'option' + option} onRemove={() => this.removeTag(option)}>
          {tagLabel}
        </Tag>
      );
    });
  };

  shouldLoadMoreResults = (): boolean => {
    const {filteredOptions, loading} = this.state;
    if (filteredOptions.length === 0 || loading) {
      return false;
    }
    this.setState({loading: true});
    return true;
  };

  loadMoreResults = () => {
    const {filteredOptions, options} = this.state;
    if (this.shouldLoadMoreResults()) {
      let currentlyLoadedOptions = options;
      let unloadedOptions = filteredOptions;

      const endIndex =
        currentlyLoadedOptions.length + this.RESULT_PAGE_LENGTH - 1;
      let optionsToLoad = unloadedOptions.slice(0, endIndex);

      currentlyLoadedOptions = currentlyLoadedOptions.concat(optionsToLoad);
      unloadedOptions = unloadedOptions.slice(
        endIndex,
        unloadedOptions.length - 1,
      );

      setTimeout(() => {
        this.setState({
          options: currentlyLoadedOptions,
          filteredOptions: unloadedOptions,
          loading: false,
        });
      }, 300);
    }
  };

  shouldShowRenderActionButton = (): boolean => {
    const {inputText} = this.state;
    return !(
      !inputText ||
      inputText === '' ||
      (this.ALLOW_MULTIPLE && containsOption(this.options, inputText))
    );
  };

  renderActionButton = (): ItemDescriptor | undefined => {
    if (this.shouldShowRenderActionButton()) {
      return {
        content: 'Add "' + this.state.inputText + '"',
        icon: 'circlePlus',
        onAction: () => this.addTag(this.state.inputText),
      };
    }
    return undefined;
  };

  getListTitle = (): string | undefined => {
    if (this.shouldShowRenderActionButton()) {
      return undefined;
    }
    return 'Frequently Used Tags';
  };

  render() {
    console.log(this.renderActionButton());

    return (
      <AppProvider>
        <Page title="Demo Tags Autocomplete">
          <br />
          <TextContainer>
            <Stack>{this.renderTags()}</Stack>
          </TextContainer>
          <br />
          <Autocomplete
            options={this.state.options}
            selected={this.state.selected}
            textField={
              <Autocomplete.ComboBox.TextField
                onChange={this.updateText}
                label=""
                value={this.state.inputText}
                placeholder="Vintage, cotton, summer"
              />
            }
            onSelect={this.updateSelection}
            allowMultiple={this.ALLOW_MULTIPLE}
            loading={this.state.loading}
            willLoadMoreResults
            onLoadMoreResults={this.loadMoreResults}
            actionBefore={this.renderActionButton()}
            listTitle={this.getListTitle()}
            onClose={this.clearText}
          />
        </Page>
      </AppProvider>
    );
  }

  filterAndUpdateOptions = (inputString: string) => {
    if (inputString === '') {
      this.setState({
        options: this.options.slice(0, this.RESULT_PAGE_LENGTH),
        filteredOptions: this.options.slice(
          this.RESULT_PAGE_LENGTH,
          this.options.length,
        ),
      });
      return;
    }

    const filterRegex = new RegExp(inputString, 'i');
    const resultOptions = this.options.filter((option) =>
      option.label.match(filterRegex),
    );
    let endIndex = resultOptions.length - 1;
    if (resultOptions.length === 0) {
      endIndex = 0;
    }
    this.setState({
      options: resultOptions.slice(0, this.RESULT_PAGE_LENGTH),
      filteredOptions: resultOptions.slice(this.RESULT_PAGE_LENGTH, endIndex),
    });
  };

  updateSelection = (updatedSelection: string[]) => {
    const selectedText = updatedSelection.map((selectedItem: string) => {
      const matchedOption = this.options.filter((option) => {
        return option.value.match(selectedItem);
      });
      return matchedOption[0] && matchedOption[0].label;
    });
    if (this.ALLOW_MULTIPLE) {
      this.setState({selected: updatedSelection});
    } else {
      this.setState({selected: selectedText, inputText: selectedText});
    }
  };
}

function containsOption(a: any, optionString: string) {
  for (var i = 0; i < a.length; i++) {
    const filterRegex = new RegExp(
      '^' + optionString.replace(' ', '_') + '$',
      'i',
    );
    if (a[i].value.match(filterRegex)) {
      return true;
    }
  }
  return false;
}

/* eslint-enable */

```

</details>

### 🎩 checklist

* [ ] Tested on mobile
* [ ] Tested on multiple browsers
* [ ] Tested for [accessibility](https://github.com/Shopify/polaris-react/blob/master/documentation/Accessibility%20testing.md)
* [ ] Updated [UNRELEASED.md](https://github.com/Shopify/polaris-react/blob/master/UNRELEASED.md)

### CODE REVIEW

File: src/components/Autocomplete/README.md [hunk 1]
```diff
@@ -80,6 +81,14 @@ class AutocompleteExample extends React.Component {
     this.filterAndUpdateOptions(newValue);
   };
 
+  clearText = () => {
+    const {inputText} = this.state;
+    if (inputText == '') {
```
Comment (AndrewMusgrave) [inline src/components/Autocomplete/README.md]: Strict equality check would be preferred over a loose equality check here however we normally go with this shorthand 

```suggestion
    if (!inputText) {
```

File: src/components/Autocomplete/README.md [hunk 2]
```diff
@@ -160,6 +171,14 @@ class MultiAutocompleteExample extends React.Component {
...
```
Comment (AndrewMusgrave) [inline src/components/Autocomplete/README.md]:
```suggestion
    if (!inputText) {
```

File: src/components/Autocomplete/components/ComboBox/tests/ComboBox.test.tsx [hunk 3]
```diff
@@ -381,6 +381,60 @@ describe('<ComboBox/>', () => {
...
+          options={options}
+          selected={[]}
+          textField={renderTextField()}
+          onSelect={noop}
+          onClose={spy}
+        />,
+      );
+
+      comboBox.simulate('blur');
+      expect(spy).toHaveBeenCalled();
```
Comment (AndrewMusgrave) [inline src/components/Autocomplete/components/ComboBox/tests/ComboBox.test.tsx]: We'll want to make sure it's only called once on blur

```suggestion
      expect(spy).toHaveBeenCalledTimes(1)
```

File: src/components/Autocomplete/tests/Autocomplete.test.tsx [hunk 4]
```diff
@@ -83,6 +84,7 @@ describe('<Autocomplete/>', () => {
         actionBefore,
       );
       expect(autocomplete.find(ComboBox).prop('onSelect')).toBe(handleOnSelect);
+      expect(autocomplete.find(ComboBox).prop('onClose')).toBe(handleOnClose);
     });
   });
```
Comment (solonaarmstrong) [inline src/components/Autocomplete/tests/Autocomplete.test.tsx]: You could also:

```suggestion
  describe('onClose', () => {
    it('calls onClose when called by ComboBox', () => {
      const spy = jest.fn();
      const autocomplete = mountWithAppProvider(
        <Autocomplete
          options={options}
          selected={[]}
          textField={renderTextField()}
          onSelect={noop}
          onClose={spy}
        />,
      );

      const comboBox = autocomplete.find(ComboBox);
      trigger(comboBox, 'onClose');
      expect(spy).toHaveBeenCalledTimes(1);
    });
  });
```

File: src/components/Autocomplete/README.md (hunk 1 above)
Comment (solonaarmstrong) [reply src/components/Autocomplete/README.md]: My bad, sorry. It should also reset the `filterAndUpdateOptions`. @darrenhebner I don't want to hijack your PR. Do you want me to make these changes or do it yourself?

```suggestion
   clearText = () => {
    const {inputText} = this.state;
    if (!inputText) {
      return;
    }
    this.setState({inputText: ''});
    this.filterAndUpdateOptions('');
  };
```

File: src/components/Autocomplete/README.md (hunk 2 above)
Comment (solonaarmstrong) [reply src/components/Autocomplete/README.md]:
```suggestion
   clearText = () => {
    const {inputText} = this.state;
    if (!inputText) {
      return;
    }
    this.setState({inputText: ''});
    this.filterAndUpdateOptions('');
  };
```

File: src/components/Autocomplete/README.md [hunk 5]
```diff
@@ -264,6 +283,14 @@ class AutocompleteExample extends React.Component {
...
+      return;
+    }
+    this.setState({inputText: ''});
+  };
+
```
Comment (solonaarmstrong) [inline src/components/Autocomplete/README.md]:
```suggestion
   clearText = () => {
    const {inputText} = this.state;
    if (!inputText) {
      return;
    }
    this.setState({inputText: ''});
    this.filterAndUpdateOptions('');
  };
```

File: src/components/Autocomplete/README.md (hunk 1 above)
Comment (darrenhebner) [reply src/components/Autocomplete/README.md]: @solonaarmstrong - Feel free to make changes to the README 👍

**Comment by darrenhebner** [comment]
Sorry for letting this sit for a bit. I'm still getting some weird behaviour with the different 'hooks' I was using to detect if the popover was closed.

I'm trying to figure out if there's a better approach to clearing the input when the merchant is done interacting with the field. `onClose` is not the right answer in all cases.

**Comment by darrenhebner** [comment]
I've pushed a commit that takes a bit of a different approach. I just want to get some feedback on it before I update the tests.

I'm having a really really hard time finding a good 'hook' for clearing the search field. What I really want is an `onBlur` that takes is called when you blur away from the whole component (wrapping combobox, text field, and popover). 

There's some unexpected behaviour with the way `handleFocus`, `handleBlur`, and `handleClick` work. All three of these events toggle the popover and all three of them are called at the same time when you click on an item in the popover. 

I've got the clearText behaviour working correctly with this hack:
https://github.com/Shopify/polaris-react/pull/479/files#diff-6d3132b11e3b2128f89557c22535505bR383

Can anyone think of another way to approach this?

**Comment by darrenhebner** [comment]
I'm going to close this PR for now. I'm unable to implement this feature (without hacks 😄) until there are some updates to the ComboBox.

The benefit to the user vs time spent adding the feature is probably not worth it for now. I have some ideas for a simpler alternative solution. I'll put a PR up shortly.

More details can be found in https://github.com/Shopify/polaris-react/issues/499

**Comment by solonaarmstrong** [comment]
New issue: https://github.com/Shopify/polaris-react/issues/576

**Pull request closed without merging**